core/src/main/java/com/digitalpebble/stormcrawler/protocol/selenium/SeleniumGridProtocol.java

```diff
@@ -0,0 +1,194 @@
+import com.digitalpebble.stormcrawler.protocol.ProtocolResponse;
+import com.digitalpebble.stormcrawler.util.ConfUtils;
+import com.google.gson.Gson;
+import java.io.IOException;
+import java.io.InputStream;
+import java.io.InputStreamReader;
+import java.io.Reader;
+import java.net.URL;
+import java.util.List;
+import java.util.Map;
+import java.util.concurrent.LinkedBlockingQueue;
+import org.apache.storm.Config;
+import org.openqa.selenium.remote.DesiredCapabilities;
+import org.openqa.selenium.remote.RemoteWebDriver;
+import org.slf4j.LoggerFactory;
+
+public abstract class SeleniumGridProtocol extends AbstractHttpProtocol {
+    protected static final org.slf4j.Logger LOG =
+            LoggerFactory.getLogger(SeleniumGridProtocol.class);
+    protected static LinkedBlockingQueue<Holder> driversQueue;
+    private NavigationFilters filters;
+    private final DesiredCapabilities capabilities = new DesiredCapabilities();
+
+    protected String gridAddress;
+
+    @Override
+    public void configure(Config conf) {
+        super.configure(conf);
+        driversQueue = new LinkedBlockingQueue<>();
+        filters = NavigationFilters.fromConf(conf);
+        gridAddress =
+                ConfUtils.getString(conf, "selenium.grid.address", "http://localhost:4444/wd/hub");
+    }
+
+    protected synchronized List<Map<String, Object>> getAllNodesList() throws IOException {
+        Map<String, Object> valueMap = null;
+        boolean ready = false;
+        while (!ready) {
+            Map<String, Object> map = getStatusStream();
+            valueMap = (Map<String, Object>) map.get("value");
+            ready = (boolean) valueMap.get("ready");
+            if (!ready) {
+                LOG.warn("Selenium Grid is not ready yet");
+            }
+        }
+        LOG.info("Grid Is Ready to Serve");
+        return (List<Map<String, Object>>) valueMap.get("nodes");
+    }
+
+    private Map<String, Object> getStatusStream() throws IOException {
+        Gson gson = new Gson();
+        URL url = new URL(gridAddress + "/status");
+        InputStream stream = url.openStream();
+        Reader reader = new InputStreamReader(stream);
+        Map<String, Object> map = gson.fromJson(reader, Map.class);
+        stream.close();
+        return map;
+    }
+
+    protected int getSessionsCount(List<Map<String, Object>> nodes) {
+        int availableSessions = 0;
+        for (Map<String, Object> node : nodes) {
+            List<Map<String, Object>> slots = (List<Map<String, Object>>) node.get("slots");
+            for (Map<String, Object> slot : slots) {
+                if (slot.get("session") == null) {
+                    availableSessions++;
+                }
+            }
+        }
+        return availableSessions;
+    }
+
+    public class Holder {
+        public RemoteWebDriver driver;
+        public Long time;
+
+        public void setDriver(RemoteWebDriver driver) {
```

Comment:
Can be removed as it is never called and driver is set via constructor.

Reply:
Since it is just a pojo class can I move it to some other place, can you please suggest a package where should I move this file. I want to avoid using the new operator inside seleniumgridimpl and directly use the Holder.java as model

Reply:
Feel free to make a suggestions.